File: dc3-common/dc3-common-public/src/main/java/io/github/pnoker/common/utils/TimeUtil.java
/*
 * Copyright 2016-present the IoT DC3 original author or authors.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

package io.github.pnoker.common.utils;

import io.github.pnoker.common.constant.common.ExceptionConstant;
import io.github.pnoker.common.constant.common.TimeConstant;
import lombok.extern.slf4j.Slf4j;

import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;
import java.time.format.DateTimeParseException;
import java.util.Calendar;
import java.util.Date;

/**
 * Legacy {@link Date} formatting and time calculation with thread-safe formatters.
 *
 * @author pnoker
 * @version 2025.9.0
 * @since 2016.10.1
 */
@Slf4j
public class TimeUtil {

    /**
     * Thread-safe date-time formatters (DateTimeFormatter is immutable and thread-safe by design).
     */
    private static final DateTimeFormatter DEFAULT_DATE_FORMATTER =
            DateTimeFormatter.ofPattern(TimeConstant.DEFAULT_DATE_FORMAT);

    private static final DateTimeFormatter COMPLETE_DATE_FORMATTER =
            DateTimeFormatter.ofPattern(TimeConstant.COMPLETE_DATE_FORMAT);

    private TimeUtil() {
        throw new IllegalStateException(ExceptionConstant.UTILITY_CLASS);
    }

    /**
     * Get milliseconds from Date
     *
     * @param date {@link Date}
     * @return Milliseconds since epoch
     */
    public static long milliSeconds(Date date) {
        return date.getTime();
    }

    /**
     * Get Date from milliseconds
     *
     * @param milliSeconds Milliseconds since epoch
     * @return Date {@link Date}
     */
    public static Date localDateTime(long milliSeconds) {
        return new Date(milliSeconds);
    }

    /**
     * Calculate expire time by adding amount to specified calendar field
     *
     * @param amount Integer amount to add
     * @param field  Calendar field : {@link Calendar Calendar.HOUR/MINUTE/...}
     * @return Date with calculated expire time
     */
    public static Date expireTime(int amount, int field) {
        Calendar calendar = Calendar.getInstance();
        calendar.setTime(new Date());
        calendar.add(field, amount);
        return calendar.getTime();
    }

    /**
     * Format date using yyyy-MM-dd HH:mm:ss pattern
     *
     * @param date {@link Date}
     * @return Formatted date string
     */
    public static String defaultFormat(Date date) {
        return date.toInstant().atZone(TimeConstant.DEFAULT_ZONEID).toLocalDateTime().format(DEFAULT_DATE_FORMATTER);
    }

    /**
     * Format date using yyyy-MM-dd HH:mm:ss.SSS pattern
     *
     * @param date {@link Date}
     * @return Formatted date string with milliseconds
     */
    public static String completeFormat(Date date) {
        return date.toInstant().atZone(TimeConstant.DEFAULT_ZONEID).toLocalDateTime().format(COMPLETE_DATE_FORMATTER);
    }

    /**
     * Parse date string yyyy-MM-dd HH:mm:ss to Date
     *
     * @param dateString Date string in yyyy-MM-dd HH:mm:ss format
     * @return Date object
     */
    public static Date defaultDate(String dateString) {
        try {
            LocalDateTime ldt = LocalDateTime.parse(dateString, DEFAULT_DATE_FORMATTER);
            return Date.from(ldt.atZone(TimeConstant.DEFAULT_ZONEID).toInstant());
        } catch (DateTimeParseException ignored) {
            log.debug("Failed to parse date string '{}' with default format", dateString);
            return null;
        }
    }

    /**
     * Parse date string yyyy-MM-dd HH:mm:ss.SSS to Date
     *
     * @param dateString Date string in yyyy-MM-dd HH:mm:ss.SSS format
     * @return Date object
     */
    public static Date completeDate(String dateString) {
        try {
            LocalDateTime ldt = LocalDateTime.parse(dateString, COMPLETE_DATE_FORMATTER);
            return Date.from(ldt.atZone(TimeConstant.DEFAULT_ZONEID).toInstant());
        } catch (DateTimeParseException ignored) {
            log.debug("Failed to parse date string '{}' with complete format", dateString);
            return null;
        }
    }

}

File: dc3-web/src/utils/timeUtil.ts
/*
 * Copyright 2016-present the IoT DC3 original author or authors.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

/**
 * Shared time-formatting helpers. Every dashboard card used to declare its
 * own `formatTime` / `humanDuration` / `formatMs`, each slightly different
 * (some dropped the date, some used toLocaleString, some toLocaleTimeString).
 * Normalised here so one change reaches every card.
 *
 * <p>Locale is hard-coded to zh-CN + 24-hour — matching the project's
 * i18n language and existing display conventions.</p>
 */

const LOCALE = 'zh-CN';

/** Parse ISO / "yyyy-MM-dd HH:mm:ss" into a Date; returns null on failure. */
export const parseDateSafe = (v: string | Date | undefined | null): Date | null => {
  if (v == null || v === '') return null;
  if (v instanceof Date) return Number.isNaN(v.getTime()) ? null : v;
  // Backend "2026-05-04 12:34:56" needs the space→T swap to round-trip via
  // the JS Date parser in every browser; ISO-8601 strings already have "T".
  const d = new Date(typeof v === 'string' && v.includes(' ') && !v.includes('T') ? v.replace(' ', 'T') : v);
  return Number.isNaN(d.getTime()) ? null : d;
};

/**
 * Full date + clock — "2026/05/04 12:34:56". Used for audit-style timestamps
 * (alarm create_time, last_seen, operate_time).
 */
export const formatDateTime = (v: string | Date | undefined | null): string => {
  const d = parseDateSafe(v);
  if (!d) return typeof v === 'string' ? v : '';
  return d.toLocaleString(LOCALE, {hour12: false});
};

/** Clock only — "12:34:56". Used by live feeds where the date context is already implicit. */
export const formatClock = (v: string | Date | undefined | null): string => {
  const d = parseDateSafe(v);
  if (!d) return typeof v === 'string' ? v : '';
  return d.toLocaleTimeString(LOCALE, {hour12: false});
};

/**
 * Month/day + clock — "05/04 12:34". Used on the timeline recent-unconfirmed
 * cards where space is tight and the year adds no info.
 */
export const formatShortDateTime = (v: string | Date | undefined | null): string => {
  const d = parseDateSafe(v);
  if (!d) return typeof v === 'string' ? v : '';
  return d.toLocaleString(LOCALE, {
    month: '2-digit',
    day: '2-digit',
    hour: '2-digit',
    minute: '2-digit',
    hour12: false,
  });
};

/**
 * Millisecond duration → concise human string: ms / s / m / h. Used for
 * MTTA display where values span 0-3600000+.
 */
export const formatMs = (ms: number | null | undefined): string => {
  if (!ms || ms <= 0) return '—';
  if (ms < 1000) return `${Math.round(ms)}ms`;
  const s = ms / 1000;
  if (s < 60) return `${s.toFixed(1)}s`;
  const m = s / 60;
  if (m < 60) return `${m.toFixed(1)}m`;
  return `${(m / 60).toFixed(1)}h`;
};

/** Seconds → compact duration: "42s" / "3m" / "2h" / "5d". */
export const humanDuration = (seconds: number | null | undefined): string => {
  if (!seconds || seconds < 0) return '—';
  if (seconds < 60) return `${Math.floor(seconds)}s`;
  if (seconds < 3600) return `${Math.floor(seconds / 60)}m`;
  if (seconds < 86400) return `${Math.floor(seconds / 3600)}h`;
  return `${Math.floor(seconds / 86400)}d`;
};
